model/riscv_vmem.sail

```diff
@@ -0,0 +1,854 @@
+  ppn_64b << pagesize_bits
+}
+
+// Compute address translation mode from SATP register
+// PRIVATE
+function translationMode(priv : Privilege) -> SATPMode = {
+  if priv == Machine then
+    Sbare
+  else if sizeof(xlen) == 32 then
+    match satp[31] {
+      bitzero => Sbare,
+      bitone  => Sv32
+    }
+  else if sizeof(xlen) == 64 then {
+    // This extend op is a no-op when xlen==64,
+    // but appeases the type-checker when xlen==32 (when this code is not executed!)
+    let satp_64b : bits(64) = zero_extend(satp);
+    // Translation mode is based on mstatus.SXL, which could be RV32 when xlen==64
+    let arch = architecture(get_mstatus_SXL(mstatus));
+    match arch {
+      Some(RV64) => { let mbits : bits(4) = satp[63 .. 60];
+                      match satp64Mode_of_bits(RV64, mbits) {    // see riscv_types.sail
+                        Some(m) => m,
+                        None()  => internal_error(__FILE__, __LINE__,
+                                                  "invalid RV64 translation mode in satp")
+                      }
+                    },
+      Some(RV32) => match satp[31] {
+                      // When xlen is 64, mstatus.SXL (for S privilege) can be RV32
+                      bitzero => Sbare,
+                      bitone  => Sv32
+                    },
+      _    => internal_error(__FILE__, __LINE__, "unsupported address translation arch")
+    }
+  }
+  else
+    internal_error(__FILE__, __LINE__, "unsupported xlen")
+}
+
+// ****************************************************************
+// VA to PA translation
+
+// Result of address translation
+
+// PUBLIC
+union TR_Result('paddr : Type, 'failure : Type) = {
+  TR_Address : ('paddr, ext_ptw),
+  TR_Failure : ('failure, ext_ptw)
+}
+
+// This function can be ignored on first reading since TLBs are not
+// part of RISC-V architecture spec (see TLB_NOTE above).
+// PRIVATE: translate on TLB hit, and maintenance of PTE in TLB
+function translate_TLB_hit(sv_params : SV_Params,
+                           asid      : bits(16),
+                           ptb       : bits(64),
+                           vAddr     : bits(64),
+                           ac        : AccessType(ext_access_type),
+                           priv      : Privilege,
+                           mxr       : bool,
+                           do_sum    : bool,
+                           ext_ptw   : ext_ptw,
+                           tlb_index : nat,
+                           ent       : TLB_Entry)
+                          -> TR_Result(bits(64), PTW_Error) = {
+  let pte       = ent.pte;
+  let ext_pte   = msbs_of_PTE(sv_params, pte);
+  let pte_lsbs  = Mk_PTE_LSBs(pte[7 .. 0]);
+  let pte_check = check_PTE_permission(ac, priv, mxr, do_sum, pte_lsbs,
+                                       ext_pte,
+                                       ext_ptw);
+  let tr_result : TR_Result(bits(64), PTW_Error)
```

Comment:
Can remove this line and just have the last expression be the `match`.

Reply:
Ok. Fixed.  I’ve commented elsewhere that sometimes naming the result is useful for debug-printing the result.